File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    const result = [];
    const [tY, tM, tD] = today.split(".").map(Number);
    const totalDay = tY*12*28 + tM*28 + tD
    
    const termMap = {};
    terms.forEach(line => {
        const [type, month] = line.split(" ");
        termMap[type] = Number(month) * 28; 
    });
    
    privacies.forEach((line, index) => {
        const [fullDate, type] = line.split(" ");
        const [y, m, d] = fullDate.split(".").map(Number);
        
        const expiryTotal = (y * 12 * 28) + (m * 28) + d + termMap[type];
        
        if (totalDay >= expiryTotal) {
            result.push(index + 1);
    };
    });
    
    
    
    

    return result
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

// "YYYY.MM.DD"
vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    // 해시 사용 
    unordered_map<char, int> t;
    for (string& s : terms) {
        char type = s[0];
        int month = stoi(s.substr(2));
        
        t[type] = month;
    }
    int todayYear = stoi(today.substr(0, 4));
    int todayMonth = stoi(today.substr(5, 2));
    int todayDay = stoi(today.substr(8, 2));
    
    int todayDays = todayYear * 12 * 28
                  + todayMonth * 28
                  + todayDay;
    
    vector<int> answer;
    
    for (int i = 0; i < privacies.size(); i++) {

        string startDate = privacies[i].substr(0, 10);
        char type = privacies[i][11];

        // 수집 날짜를 일수로 변환
        int year = stoi(startDate.substr(0, 4));
        int month = stoi(startDate.substr(5, 2));
        int day = stoi(startDate.substr(8, 2));

        int startDays = year * 12 * 28
                      + month * 28
                      + day;
        
        int expiration = startDays + t[type] * 28;
        
        if (expiration <= todayDays) {
            answer.push_back(i + 1);
        }
    }



    return answer;
}